src/Plugin/EntityReferenceSelection/OgSelection.php

```diff
@@ -77,37 +151,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
       $query->condition($bundle_key, $bundles, 'IN');
     }
 
-    $user_groups = $this->getUserGroups();
-    if (!$user_groups) {
+    // Get the identifier key of the entity.
+    $identifier_key = $definition->getKey('id');
+
+    if ($this->currentUser->isAnonymous()) {
+      // @todo: Check if anonymous users should have access to any referenced
+      // groups? What about groups that allow anonymous posts?
+      return $query->condition($identifier_key, -1);
+    }
+
+    if ($this->currentUser->hasPermission('administer group')) {
+      // User can see all the groups.
       return $query;
     }
 
-    $identifier_key = $definition->getKey('id');
+    if (empty($this->configuration['entity'])) {
+      // @todo: Find out why we have this scenario.
+      return $query;
+    }
+
+    /** @var \Drupal\Core\Entity\EntityInterface $entity */
+    $entity = $this->configuration['entity'];
+    $entity_type_id = $entity->getEntityTypeId();
+    $bundle = $entity->bundle();
 
     $ids = [];
-    if (!empty($this->configuration['handler_settings']['field_mode']) && $this->configuration['handler_settings']['field_mode'] == 'admin') {
-      // Don't include the groups, the user doesn't have create permission.
-      foreach ($user_groups as $delta => $group) {
+    foreach ($this->getUserGroups() as $group) {
+      // Check user has "create" permission on this entity.
+      if ($this->ogAccess->userAccess($group, "create $entity_type_id $bundle", $this->currentUser)->isAllowed()) {
```

Comment:
We should fix this in #510:
```suggestion
      $permission = "create $bundle content";
      if ($this->ogAccess->userAccess($group, $permission, $this->currentUser)->isAllowed()) {
```